File: sdk/typescript/src/features/__tests__/rule-engine.test.ts
import RuleEngine from '../rule-engine';

// Mock OpenlitConfig
jest.mock('../../config', () => ({
  __esModule: true,
  default: {
    applicationName: 'test-app',
    environment: 'test',
  },
}));

// Mock constant
jest.mock('../../constant', () => ({
  OPENLIT_URL: 'http://127.0.0.1:3000',
}));

// Save original env
const originalEnv = process.env;

beforeEach(() => {
  jest.restoreAllMocks();
  process.env = { ...originalEnv };
  delete process.env.OPENLIT_URL;
  delete process.env.OPENLIT_API_KEY;
});

afterAll(() => {
  process.env = originalEnv;
});

describe('RuleEngine.evaluate', () => {
  it('sends correct request to rule-engine evaluate endpoint', async () => {
    const mockResponse = {
      matchingRuleIds: ['rule-1'],
      entities: [{ rule_id: 'rule-1', entity_type: 'context', entity_id: 'ctx-1' }],
    };

    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve(mockResponse),
    });

    const result = await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      apiKey: 'test-key',
      entityType: 'context',
      fields: { 'gen_ai.system': 'openai' },
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const [url, options] = (global.fetch as jest.Mock).mock.calls[0];
    expect(url).toBe('http://localhost:3000/api/rule-engine/evaluate');
    expect(options.method).toBe('POST');
    expect(options.headers.Authorization).toBe('Bearer test-key');

    const body = JSON.parse(options.body);
    expect(body.entity_type).toBe('context');
    expect(body.fields).toEqual({ 'gen_ai.system': 'openai' });
    expect(body.source).toBe('ts-sdk');

    expect(result).toEqual(mockResponse);
  });

  it('returns error object on HTTP error', async () => {
    global.fetch = jest.fn().mockResolvedValue({
      ok: false,
      status: 401,
    });

    const result = await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      apiKey: 'bad-key',
      entityType: 'context',
      fields: {},
    });

    expect(result).toHaveProperty('err');
    expect((result as any).err).toContain('401');
  });

  it('returns error object on network failure', async () => {
    global.fetch = jest.fn().mockRejectedValue(new Error('Network error'));

    const result = await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      apiKey: 'test-key',
      entityType: 'context',
      fields: {},
    });

    expect(result).toHaveProperty('err');
    expect((result as any).err).toContain('Network error');
  });

  it('uses OPENLIT_URL env var as fallback', async () => {
    process.env.OPENLIT_URL = 'http://env-host:4000';

    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ matchingRuleIds: [] }),
    });

    await RuleEngine.evaluate({
      apiKey: 'key',
      entityType: 'context',
      fields: {},
    });

    const [url] = (global.fetch as jest.Mock).mock.calls[0];
    expect(url).toBe('http://env-host:4000/api/rule-engine/evaluate');
  });

  it('uses OPENLIT_API_KEY env var as fallback', async () => {
    process.env.OPENLIT_API_KEY = 'env-api-key';

    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ matchingRuleIds: [] }),
    });

    await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      entityType: 'context',
      fields: {},
    });

    const [, options] = (global.fetch as jest.Mock).mock.calls[0];
    expect(options.headers.Authorization).toBe('Bearer env-api-key');
  });

  it('passes includeEntityData and entityInputs', async () => {
    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ matchingRuleIds: [], entity_data: {} }),
    });

    await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      apiKey: 'key',
      entityType: 'prompt',
      fields: { key: 'val' },
      includeEntityData: true,
      entityInputs: { variables: { name: 'test' } },
    });

    const body = JSON.parse((global.fetch as jest.Mock).mock.calls[0][1].body);
    expect(body.include_entity_data).toBe(true);
    expect(body.entity_inputs).toEqual({ variables: { name: 'test' } });
    expect(body.entity_type).toBe('prompt');
  });

  it('includes metaProperties from OpenlitConfig', async () => {
    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ matchingRuleIds: [] }),
    });

    await RuleEngine.evaluate({
      url: 'http://localhost:3000',
      apiKey: 'key',
      entityType: 'context',
      fields: {},
    });

    const body = JSON.parse((global.fetch as jest.Mock).mock.calls[0][1].body);
    expect(body.metaProperties).toEqual({
      applicationName: 'test-app',
      environment: 'test',
    });
  });
});

File: sdk/go/rule_engine_test.go
package openlit

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
	"time"
)

func TestEvaluateRule_Success(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/rule-engine/evaluate" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
		if r.Method != http.MethodPost {
			t.Errorf("unexpected method: %s", r.Method)
		}
		if r.Header.Get("Authorization") != "Bearer test-key" {
			t.Errorf("unexpected auth header: %s", r.Header.Get("Authorization"))
		}
		if r.Header.Get("Content-Type") != "application/json" {
			t.Errorf("unexpected content-type: %s", r.Header.Get("Content-Type"))
		}

		// Decode and verify request body
		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("failed to decode request body: %v", err)
		}
		if body["entity_type"] != "context" {
			t.Errorf("expected entity_type=context, got %v", body["entity_type"])
		}
		if body["source"] != "go-sdk" {
			t.Errorf("expected source=go-sdk, got %v", body["source"])
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"matchingRuleIds": []string{"rule-1"},
			"entities": []map[string]string{
				{"rule_id": "rule-1", "entity_type": "context", "entity_id": "ctx-1"},
			},
		})
	}))
	defer srv.Close()

	result, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		URL:        srv.URL,
		APIKey:     "test-key",
		EntityType: RuleEntityContext,
		Fields: map[string]interface{}{
			"gen_ai.system": "openai",
		},
	})

	if err != nil {
		t.Fatalf("EvaluateRule: %v", err)
	}
	if len(result.MatchingRuleIDs) != 1 || result.MatchingRuleIDs[0] != "rule-1" {
		t.Errorf("unexpected matchingRuleIds: %v", result.MatchingRuleIDs)
	}
	if len(result.Entities) != 1 || result.Entities[0].EntityID != "ctx-1" {
		t.Errorf("unexpected entities: %v", result.Entities)
	}
}

func TestEvaluateRule_WithEntityData(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body map[string]interface{}
		json.NewDecoder(r.Body).Decode(&body)

		if body["include_entity_data"] != true {
			t.Errorf("expected include_entity_data=true, got %v", body["include_entity_data"])
		}
		if body["entity_type"] != string(RuleEntityPrompt) {
			t.Errorf("expected entity_type=prompt, got %v", body["entity_type"])
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"matchingRuleIds": []string{},
			"entities":        []interface{}{},
			"entity_data":     map[string]interface{}{"prompt": "Hello"},
		})
	}))
	defer srv.Close()

	result, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		URL:               srv.URL,
		APIKey:            "test-key",
		EntityType:        RuleEntityPrompt,
		Fields:            map[string]interface{}{"key": "val"},
		IncludeEntityData: true,
		EntityInputs:      map[string]interface{}{"variables": map[string]string{"name": "test"}},
	})

	if err != nil {
		t.Fatalf("EvaluateRule: %v", err)
	}
	if result.EntityData == nil {
		t.Fatal("expected entity_data to be present")
	}
	if result.EntityData["prompt"] != "Hello" {
		t.Errorf("unexpected entity_data: %v", result.EntityData)
	}
}

func TestEvaluateRule_MissingAPIKey(t *testing.T) {
	os.Unsetenv("OPENLIT_API_KEY")

	_, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		URL:        "http://localhost:3000",
		EntityType: RuleEntityContext,
		Fields:     map[string]interface{}{},
	})

	if err == nil {
		t.Fatal("expected error for missing API key")
	}
	if !strings.Contains(err.Error(), "missing API key") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestEvaluateRule_ServerError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
	}))
	defer srv.Close()

	_, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		URL:        srv.URL,
		APIKey:     "test-key",
		EntityType: RuleEntityContext,
		Fields:     map[string]interface{}{},
	})

	if err == nil {
		t.Fatal("expected error for 500 response")
	}
	if !strings.Contains(err.Error(), "500") {
		t.Errorf("expected 500 in error, got: %v", err)
	}
}

func TestEvaluateRule_EnvVarFallback(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer env-key" {
			t.Errorf("expected env API key, got: %s", r.Header.Get("Authorization"))
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"matchingRuleIds": []string{},
			"entities":        []interface{}{},
		})
	}))
	defer srv.Close()

	os.Setenv("OPENLIT_URL", srv.URL)
	os.Setenv("OPENLIT_API_KEY", "env-key")
	defer os.Unsetenv("OPENLIT_URL")
	defer os.Unsetenv("OPENLIT_API_KEY")

	result, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		EntityType: RuleEntityContext,
		Fields:     map[string]interface{}{"key": "val"},
	})

	if err != nil {
		t.Fatalf("EvaluateRule: %v", err)
	}
	if result == nil {
		t.Fatal("expected non-nil result")
	}
}

func TestEvaluateRule_CustomTimeout(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(200 * time.Millisecond)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"matchingRuleIds": []string{},
			"entities":        []interface{}{},
		})
	}))
	defer srv.Close()

	// Very short timeout should fail
	_, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		URL:        srv.URL,
		APIKey:     "test-key",
		EntityType: RuleEntityContext,
		Fields:     map[string]interface{}{},
		Timeout:    1 * time.Millisecond,
	})

	if err == nil {
		t.Fatal("expected timeout error")
	}
}

func TestEvaluateRule_DefaultURL(t *testing.T) {
	os.Unsetenv("OPENLIT_URL")

	// This will fail to connect but we can check the error contains the default URL
	_, err := EvaluateRule(context.Background(), EvaluateRuleOptions{
		APIKey:     "test-key",
		EntityType: RuleEntityContext,
		Fields:     map[string]interface{}{},
		Timeout:    100 * time.Millisecond,
	})

	// Should fail with connection refused (default localhost:3000)
	if err == nil {
		t.Fatal("expected connection error to default URL")
	}
}

func TestRuleEntityType_Constants(t *testing.T) {
	if RuleEntityContext != "context" {
		t.Errorf("RuleEntityContext = %q, want %q", RuleEntityContext, "context")
	}
	if RuleEntityPrompt != "prompt" {
		t.Errorf("RuleEntityPrompt = %q, want %q", RuleEntityPrompt, "prompt")
	}
	if RuleEntityEvaluation != "evaluation" {
		t.Errorf("RuleEntityEvaluation = %q, want %q", RuleEntityEvaluation, "evaluation")
	}
}
